### ISSUE TITLE
Fine-grained notifications are broken when applied on a synchronized Realm collection

### ISSUE THREAD

**Issue by Taco55** [realm/realm-cocoa#4800]
I have some problems with the fine-grained notifications on a Realm collection in a synchronized Realm. Although I am able to apply the notifications successfully for various collections, I am not able to get it fully working on a specific collection in my main application. 

The problem I encounter is that the notification block is sometimes not triggered after inserting, deleting or modifying an object in the corresponding Realm collection (neither an .initial or .update occurs). Unfortunately, I have no clue in what exact situations the triggers remains absent, but it seems that it mostly occur for inserts. 

What surprises me, however, is that when the same notification block is applied by another user with permission to the same Realm, the triggers are ALWAYS provided correctly for that user! Furthermore, when this other user applies changes that affect the collection, the triggers are also  delivered correctly to both users. Therefore, the notification block is applied correctly.

Because of this discrepancy, I was wondering whether the run loop might have been blocked by another activity. But as far I could detect, this is not the case since the only action performed is the action that should trigger the notification block. 
Furthermore, there are no retain cycles of previous notification tokens.

I find it hard to get the described situation reproduced in another case. Therefore, I was wondering whether there might be other cases known that could block a notification from delivering.

I would really appreciate if someone could point me in the right direction to find the underlaying cause of this problem.

**Comment by Taco55** [comment]
I just found out, by accident, that when I disable ROS (I run the MacOS version during development) and the user is still working on the InMemory Realm, all notifications are delivered correctly again. How would this be possible? Is there any known situation that ROS may block fine-grained notifications?

**Comment by istx25** [comment]
Hi @Taco55. Thanks for reaching out about this. Would you be able to share a sample project that demonstrates the issue that you're experiencing? From there, we'll have more to go on and be able to better diagnose the issue.

**Comment by Taco55** [comment]
I have created a sample app to illustrate the problem. 

Since notification are only missed for the user that owns the Realm, I implemented an escape method using a timer to check whether the notification block was triggered, and if not, by updating the data after the timer was fired.

Although I hope that I made a mistake with my implementations, it really seems a very annoying bug in the fine-grained notifications, which - in my opinion - is used for a basic use case here. 

I also checked the demo drawing app, but what surprised me is that it is based on a single user only. Have the fine-grained notifications also been tested when multiple users work on the same Realm? 

A complete sample app can be downloaded [here](https://github.com/Taco55/NotificationDemo) and should run directly. 
Unfortunately, it is not possible with ROS to determine whether a user is already registered or not. Therefore, I save the register status in `NSUserDefaults`. However, this status may become out of sync with the "true" register status of ROS. Therefore, I assume that for every test case both `NSUserDefaults` and ROS are reset. 

When the app is used on multiple devices, it can easily be observed that notifications are always delivered correctly for user2, while a significant number of notifications are missed when user1 (who shares the Realm) modifies the relevant Realm collection. 

Please let me know whether you are able to reproduce the issue with this app.

**Comment by istx25** [comment]
Thanks for sending that sample app over. I'll get someone to review what you've shared and then they'll follow-up with you about what they've found.

**Comment by Taco55** [comment]
Would it be possible to confirm (or disprove) that this is a bug? Currently, it's a blocking issue to me for further development wit ROS.
Don't understand me wrong, I really hope that I made the mistake here, but it seems a significant flaw in the implementation of the fine-grained notifications that makes bi-directional syncing hard to establish.

**Comment by jpsim** [comment]
I'm sorry @Taco55 I know you want an answer as soon as possible, but even with the repro case you shared (which definitely helps a lot!) it still takes time to get everything to a state where we can reproduce it and investigate. What doesn't help is that you didn't share specific steps to reproduce, and even then there's a considerable amount of code involved that we need to go through everything very carefully to see if you're making some incorrect assumptions in your app, or if there's a problem with Realm.

**Comment by jpsim** [comment]
So I got the app running on two devices (1 simulator, 1 physical iPhone), connecting to an instance of the Realm Object Server running locally on my mac, and at this point I'm randomly trying things in the app (since I couldn't find repro steps, if you provided them), and I'm running more into assumptions about how the sample is supposed to be used rather than issues with Realm.

For example:

```swift
dog.name = String(format: "%@ - added by %@", text, currentAppUser!.username!)
//                                                                ^ unexpectedly found nil while unwrapping an Optional value
```

So rather than continue with this, I'll wait for you to share exact steps we can take to reproduce the problem you're seeing and get back to it.

**Comment by Taco55** [comment]
I completely understand how busy you are and that unraveling this issue may also take considerable time. So, I certainly didn't expect a quick solution. I just hoped to hear whether this might have been a known issue or that this was a use case that should just work. Anyway, it would  probably still require you to investigate it in more detail, so I will patiently wait for that! 

Nevertheless, since the bi-directional syncing would hardly possible to use correctly for me for this use case, any eyes on this is really appreciated.

I am sorry that I didn't give more detailed information in advance, therefore I will provide them now. I also updated the app itself somewhat so please  use the improved demo app on my [repository](https://github.com/Taco55/NotificationDemo).

Since it is just a demo I didn't implement a complete error handling. However, when the flow is followed as described below, the issue should become apparent. In fact, the main problem is at line 244 in `ViewController.swift` where a fine-grained notification is defined and does not get triggered in the situation described.

One more note before starting the app. Please start with empty `UserDefaults` (i.e. remove app before running) and assure that user1 and user2 are not yet registered in ROS. Because there is no API method to determine whether a user is already registered (except by catching a login error) I save this status in `UserDefaults`, but this is just a quick work-around does not work ideal.

The steps that are performed automatically at first launch:
- When app starts user2 is registered (and seeded with a dog named user2_dog)
- User2 is logged out and user1 is registered (and seeded with a dog named user1_dog)
- User1 grants permission of his Realm to user 2 and logs out
- User2 logs in again to stores the identity of user1 to get track of the shared Realm Path (this path is stored the AppUser Object 
- User2 logs out and the landing view controller (ViewController.swift) becomes the root view controller
- When view did load has been triggered, user1 is logged showing a single user1_dog in a tableview
- A fine-grained notification is applied to a collection of Dog objects of user1  

At this moment user1 and user2 are registered and user2 has permission to the Realm of user1. This should be indicated by the output on the console:

```
New user user2 registered: seed with a dog
New user user1 registered: seed with a dog
Add identity of user 1 to AppUser object to Realm of user 2 (i.e. to get track of the shared Realm owned by user 1
```

The problem could already be observed at this point: when a dog is added to the Realm of user1 with the controls in the view controller, the notification block does not get triggers after the collection change. A timer that is implemented fires after 1 second after the dog is added and checks whether the notification has been triggered. When this has not been occurred within 1 second  the following output is printed:

`Escape by timer: notification is not triggered (despite a collection change). Tableview is updated anyway...`

Although, this already illustrate the problem, it can also be observed that when user2 adds a dog to this Realm (on a different device ), the notification block is triggered correctly!!! 

To reproduce this, the app should run on a physical device. However, since the `UserDefaults` on that device will be empty at first launch (indicating that the users have not yet been registered), , results in a ROS error indicating that user are already registered. After the app is run a second time, the `UserDefaults` have been updated and should launch now. I known that this is a terrible work-around (but at least a quick one ;-), but hopefully a client API method becomes available to check the register status in a better way! (see also my feature request [#142](https://github.com/realm/realm-mobile-platform/issues/142) in the ROS repo).

When you manually login user2 by clicking on the right barbutton and adds a dog, it will be added to the Realm of user1 and with the notification both the device and simulator are updated as expected! Note, however, that the debugger output is on the physical device of course...


Hopefully the problem could be reproduced with these steps and many thanks in advance.

**Comment by austinzheng** [comment]
Thanks for providing updated reproduction instructions. We'll take a closer look at your project next week and figure out how to proceed from there.

**Comment by TimOliver** [comment]
Hi @Taco55! Sorry for the delay! Thanks a lot for sending us that sample app. I've been testing it out and can reproduce the issue you've described. Curiously enough, I've discovered that deleting objects will trigger the notification block; it's just that adding them isn't working.

I'll keep stepping through it to try and identify the issue and will let you know if I find anything. Thanks!

**Comment by Taco55** [comment]
Thanks for testing the demo-app. Although the issue is quite nasty, I am glad that you could reproduce it. Since  bi-directional syncing is almost impossible to implement, I tried numerous ways to get it working before I considered that the problem could be in the Realm notifications, instead of my own code.

It is correct that the issue only occurs for inserts or modifications. Deletes are usually triggered correctly. In addition, when the Realm object server is force-quitted while keeping the in-memory Realm active, all notifications are triggered correctly! Curious…

Furthermore, a closely related (and also curious) phenomenon is that notifications are NEVER triggered in case a second in-memory Realm is active. But this only happens when this Realm is owned by the current `SyncUser` AND there is an active connection to ROS. In ALL other cases (e.g. shared Realms, disconnected ROS), notifications are delivered correctly (I already mentioned this issue shortly at [#142](https://github.com/realm/realm-mobile-platform/issues/142)). 

Let me know if I can be of help in further identifying the underlying issue.

**Comment by TimOliver** [comment]
Hi @Taco55! Not at all! If this is a legitimate Realm bug, then we're all very invested in making sure we get to the bottom of it.

Okay, so I spent a few hours today whittling back the code. I got it back as far as even removing all of the permissions, and even user 2 entirely, and the issue still persists. I've gone through all of the logic with as much detail as I possibly could, and I THINK I haven't missed anything obvious.

I've attached my modified version of your original sample app here -> [NotificationDemo.zip](https://github.com/realm/realm-cocoa/files/923341/NotificationDemo.zip)

Can you please test this out on your end and confirm that this is the same issue you're seeing?

The steps to set this project are:
1. Run a local instance of the Realm Object Server on your Mac. (This app connects to `localhost` as the URLs for the auth and Realm server).
2. Manually create a user named `user1@realm.io` and password `user1` in the Realm Object Server dashboard.
3. Install Realm via CocoaPods, and then run the app in the iOS Simulator.
4. Attempt to add a new Dog to the table.

Can you please confirm that this is the same issue you were experiencing? If so, I'll escalate it to one of Realm's engineers who worked on the notification system so they can backtrack what's going on here internally.

Thanks a lot! Have a nice weekend!

**Comment by Taco55** [comment]
Yes, this is the same issue. A manually created user and an active connection with ROS is indeed sufficient to reproduce it (I only had to enable `loginUser1Button` in the modified app before launching).

Thanks for further escalating this bug. Hopefully, it won’t be too harsh ;-)

**Comment by TimOliver** [comment]
Quick update. I tried updating to Realm 2.6.1 and integrating the new `Realm.asyncOpen` API. I was able to add one `Dog` to the list and have the notification block trigger successfully, but then it started failing on subsequent adds again. :/

We'll keep looking into it!

**Comment by jpsim** [comment]
This should have been fixed by realm/realm-object-store#433, which was merged to `master` in this repo in #4885.

@Taco55 could you please try with the latest `master` of this repo to confirm that fixes the issue for you?

**Comment by TimOliver** [comment]
Awesome! Thanks so much for that @jpsim! I'll test it in my version of the sample app and see if that fixes it on my end. :)

**Comment by TimOliver** [comment]
Drat. The problem is still there for me, even against the latest version of `master`.

@Taco55 Just in case, can you please test yours as well and see if you're having the same issue?

**Comment by jpsim** [comment]
@TimOliver that explains why you were still seeing the issue. You'll only build the Swift files from source, whereas it's Realm Objective-C that has the fix. Add this to your Podfile:

```ruby
pod 'Realm', git: 'https://github.com/realm/realm-cocoa.git', branch: 'master', submodules: true
pod 'RealmSwift', git: 'https://github.com/realm/realm-cocoa.git', branch: 'master', submodules: true
```

**Comment by TimOliver** [comment]
Ah. Whoops. Thanks for that @jpsim! I thought CocoaPods would have taken care of that automatically.

Woohoo! That did it! It's working in my sample app now!

@Taco55 Can you try pulling in the latest version of Realm using JP's `Podfile` configuration and confirm it's working for you too?

**Comment by Taco55** [comment]
@jpsim, @TimOliver Many thanks for providing a fix so quickly after you reproduced the issue.
Just tried it on the sample app and insertions indeed trigger notifications now! Awesome!

However, there is still a small remaining issue. When there are more than three dogs added and a dog is removed, not being the last or second last, the `RealmCollectionChange` instance indicates two deletes and one insert. In contrast, when the last or second last dog is deleted, the `RealmCollectionChange` instance correctly indicates one delete. 

Although I can definitely live with this issue, I can image that you want to have this working correctly  as well.

At a later time, I will try whether this fix also solves the related issue with the notifications with multiple in-memory Realms (as mentioned in a post above). I'll let you know.

**Comment by jpsim** [comment]
> However, there is still a small remaining issue. When there are more than three dogs added and a dog is removed, not being the last or second last, the `RealmCollectionChange` instance indicates two deletes and one insert. In contrast, when the last or second last dog is deleted, the `RealmCollectionChange` instance correctly indicates one delete.

Ah, that's a new one. Could you please file a new issue, hopefully with another small self-contained repro case for that too?

**Comment by tgoyne** [comment]
Objects in a Realm are not ordered, and deleting an object is one of the things that can change their order. The notification reports this as a deletion at the old location and an insertion at the new location because support for reporting it as a move is incomplete.

**Comment by jpsim** [comment]
Oh if this is for an unsorted `Results`, then yes. That's correct.

In any case, I think the core issue tracked here has been resolved, so I'll be closing this, but please do file new issues for further problems you encounter.

**Comment by Taco55** [comment]
Yes, a good point!  I just sorted the `Results` collection by date created and the notifications are now correct!
I didn't realize this before since the actual "order" didn't change (and I don't know how the underlying object store handles a delete exactly). Anyway, just ordering a collection suffices... 

Thanks again for the fix!
